### PR TITLE
Automated cherry pick of #62172: When using custom network with IP-alias, use the former's #62193: Fix IP-alias subnet creation logic #62398: Fix subnet cleanup logic when using IP-aliases with custom

### PR DESCRIPTION
**Pull request by shyamjvs**
Cherry pick of #62172 #62193 #62398 on release-1.8.

#62172: When using custom network with IP-alias, use the former's
#62193: Fix IP-alias subnet creation logic
#62398: Fix subnet cleanup logic when using IP-aliases with custom

```release-note
Add support in GCE setup scripts to create cluster with custom network. Also bug-fixes to allow using custom network with IP-aliases.
```

### CODE REVIEW

**Comment by BenTheElder** [comment]
cc @krzyzacy 
This still needs cherry pick approval.

**Comment by BenTheElder** [comment]
> I0420 00:17:29.371] Creating subnet e2e-86992-0a3d25:IP_ALIAS_SUBNETWORK
W0420 00:17:30.014] ..ERROR: (gcloud.beta.compute.networks.subnets.create) Could not fetch resource:
W0420 00:17:30.015]  - Invalid value for field 'resource.name': 'IP_ALIAS_SUBNETWORK'. Must be a match of regex '(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)'

:/
/retest

**Comment by BenTheElder** [comment]
@shyamjvs I don't think this quite fixes the bug for 1.8. This is a problem as we have to use stale gcloud, and in the future we will need to upgrade gcloud in order to switch the release scripts to the non-deprecated docker auth solution..

**Comment by shyamjvs** [comment]
@kubernetes/test-infra-maintainers The kubemark presubmit is failing while cluster Up due to:

```
W0423 10:16:32.381] ..ERROR: (gcloud.beta.compute.networks.subnets.create) Could not fetch resource:
W0423 10:16:32.381]  - Invalid value for field 'resource.name': 'IP_ALIAS_SUBNETWORK'. Must be a match of regex '(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)'
```

Seems like instead of taking the value of the IP_ALIAS_SUBNETWORK env var, it's taking `IP_ALIAS_SUBNETWORK` itself as the value.
It's coming from this line:

https://github.com/shyamjvs/kubernetes/blob/b2e00f95e812372845e72176f73b21afa7c96321/cluster/gce/config-default.sh#L237

I think the problem is with bash version(?) as the same change works on 1.9 (https://github.com/kubernetes/kubernetes/pull/62851) and 1.10 (https://github.com/kubernetes/kubernetes/pull/62847).

**Comment by BenTheElder** [comment]
@jpbetz without this bug fixed it's not possible to create a GCE cluster using recent gcloud. for CI we are working around by specially pinning some 1.8 tasks to an old image which happens to have an old enough gcloud command line but we really should fix this :(

**Comment by BenTheElder** [comment]
Er better background here from @shyamjvs for 1.9: https://github.com/kubernetes/kubernetes/pull/62851#issuecomment-383747892

**Comment by wojtek-t** [comment]
@shyamjvs - please fix ^^^

**Comment by shyamjvs** [comment]
Thanks Jeff.
Wojtek - I'm currently doing that.

**Comment by shyamjvs** [comment]
> Joe Betz without this bug fixed it's not possible to create a GCE cluster using recent gcloud. for CI we are working around by specially pinning

@BenTheElder I don't think this change is related to gcloud version. This is about (custom network + IP-alias) combination not working in our setup scripts and hence failing the kubemark presubmit which is using it.

**Comment by shyamjvs** [comment]
I solved the problem by adding support for custom-network (which wasn't there earlier on 1.8 and hence SUBNETWORK and CREATE_CUSTOM_NETWORK variables were undefined).

There is still this problem of using 'IP_ALIAS_SUBNETWORK' (that Jeff pointed out above), but we should never use it given that SUBNETWORK should always be set when CREATE_CUSTOM_NETWORK=true. I can fix that and create cherrypicks on other releases but IMO it isn't worth it (unless someone disagrees).

File: cluster/gce/config-default.sh [hunk 1]
```diff
@@ -231,11 +238,13 @@ if [ ${ENABLE_IP_ALIASES} = true ]; then
   # Size of ranges allocated to each node. Currently supports only /32 and /24.
   IP_ALIAS_SIZE=${KUBE_GCE_IP_ALIAS_SIZE:-/24}
   IP_ALIAS_SUBNETWORK=${KUBE_GCE_IP_ALIAS_SUBNETWORK:-${INSTANCE_PREFIX}-subnet-default}
+  # If we're using custom network, use the subnet we already create for it as the one for ip-alias.
+  # Note that this means SUBNETWORK would override KUBE_GCE_IP_ALIAS_SUBNETWORK in case of custom network.
+  if [[ "${CREATE_CUSTOM_NETWORK}" == true ]]; then
+    IP_ALIAS_SUBNETWORK="${SUBNETWORK:-IP_ALIAS_SUBNETWORK}"
```
Comment (wojtek-t) [inline cluster/gce/config-default.sh]: not fixed here

File: cluster/gce/config-test.sh [hunk 2]
```diff
@@ -271,11 +275,13 @@ if [ ${ENABLE_IP_ALIASES} = true ]; then
   # Size of ranges allocated to each node. gcloud current supports only /32 and /24.
   IP_ALIAS_SIZE=${KUBE_GCE_IP_ALIAS_SIZE:-/24}
   IP_ALIAS_SUBNETWORK=${KUBE_GCE_IP_ALIAS_SUBNETWORK:-${INSTANCE_PREFIX}-subnet-default}
+  # If we're using custom network, use the subnet we already create for it as the one for ip-alias.
+  # Note that this means SUBNETWORK would override KUBE_GCE_IP_ALIAS_SUBNETWORK in case of custom network.
+  if [[ "${CREATE_CUSTOM_NETWORK}" == true ]]; then
+    IP_ALIAS_SUBNETWORK="${SUBNETWORK:-IP_ALIAS_SUBNETWORK}"
```
Comment (wojtek-t) [inline cluster/gce/config-test.sh]: not fixed here

**Comment by wojtek-t** [comment]
> I solved the problem by adding support for custom-network (which wasn't there earlier on 1.8 and hence SUBNETWORK and CREATE_CUSTOM_NETWORK variables were undefined).

Ok - thanks.

> There is still this problem of using 'IP_ALIAS_SUBNETWORK' (that Jeff pointed out above), but we should never use it given that SUBNETWORK should always be set when CREATE_CUSTOM_NETWORK=true. I can fix that and create cherrypicks on other releases but IMO it isn't worth it (unless someone disagrees).

I disagree. We should fix that at head and cherrypick. But it can be separate PR + separate set of cherrypicks.

/lgtm

File: cluster/gce/config-default.sh (hunk 1 above)
Comment (shyamjvs) [reply cluster/gce/config-default.sh]: Correct... I explained why I left that unfixed in https://github.com/kubernetes/kubernetes/pull/62854#issuecomment-383906900
I can do that if and cherrypick elsewhere if you think it;s worth it.

**Comment by shyamjvs** [comment]
SGTM

**Comment by shyamjvs** [comment]
I created https://github.com/kubernetes/kubernetes/pull/63074 making that fix on HEAD and cherry-picked that commit into this PR (to save creating another one)

**Comment by wojtek-t** [comment]
kubemark is still failing...

**Comment by shyamjvs** [comment]
Hopefully it works now.

**Comment by shyamjvs** [comment]
And... kubemark is green.
Could you lgtm?

**Comment by shyamjvs** [comment]
The other failures seem unrelated.
/retest

**Comment by shyamjvs** [comment]
/hold cancel
Cancelling hold (ref https://github.com/kubernetes/kubernetes/pull/62854#issuecomment-383749950) as 1.8.12 has been cut today morning.
Also marking this as queue-fix since this is blocking PRs on 1.8 due to kubemark failures.

**Comment by BenTheElder** [comment]
grrrr, DNS flakes again :(
/retest

**Pull request merged**